### PR TITLE
[TwigComponent] Add support for anonymous component name resolution in other bundle

### LINKED ISSUE #1998: [Feature Request]  Anonymous Name Resolution Fallback
It would be cool, if the Anonymous component with the same name as its Directory would be used as a fallback if one with an exact match isn't found

Currently, you would do:
`/components/Menu/Menu.html.twig`
`<twig:Menu:Menu/>`

It would be cool to be able to write:
`<twig:Menu/>`
And have it fallback to
`/components/Menu/Menu.html.twig`
in case that:
`/components/Menu.html.twig`
is not found.

**Reasoning**
| Pros ✅                                                                        | Cons  ❌                                                                     |
|---------------------------------------------------------------------|---------------------------------------------------------------------|
| If Components get more complex, you tend to split them up, and organize them in subdirectories. Having the "Index" component outside its subdir only makes it more cluttered and harder to find stuff.  This would fix that. | This can only be a fallback, to ensure backwards compatibility and because of Name conflicts |
|You Don't Get Duplication in Names anymore, which make it easier to read, and more fun to write templates. |

### PR DESCRIPTION
**Pull request by Phil23**
| Q              | A
| -------------- | ---
| Bug fix?       | no
| New feature?   | yes
| Deprecations?  | no
| Documentation? | no
| Issues         | Fix #1998
| License        | MIT

In a recent PR (#3169), @andyexeter added support for using `index.html.twig` as a component template when the component is placed in a directory named after itself. However, it seems to miss the implementation for external bundles. This PR extends the change so that anonymous component templates in other bundles can also use this naming pattern.

### CODE REVIEW

**Comment by smnandre** [comment]
Hi @Phil23 

Could you add a few tests for the edge cases... that will mainly be used to decide once and keep the same behaviour afterward. 

```twig
<twig:Acme:Foo:Index /> 

<twig:Acme:Foo:index />

<twig:Acme:Index:index />

```

( depending on the various possibilities (there is or not a template, caps / lower) )

Also let's add a line in the CHANGELOG (the one in the src/TwigComponent directory)


Minor/personal note here: this will work (and thank you for this PR by the way 🤝 ).. 

... I would still not recommand "officially" this trick to bundle maintainers.. or I would recommand them to not abuse it, and keep it "inside" their component structure instead of using this for the "main ones" -- button for instance.. 
... as the differences with how the Twig and Live component are handled could lead to frustrations I guess.

**Comment by Phil23** [comment]
@smnandre Thank you for the feedback.

I added several edge cases to the unit test.

I also added the changelog entry. I wasn't sure about the version number, though.

File: src/TwigComponent/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # CHANGELOG
 
+## 2.33
+
+-  Extended the support for using the directory name as a component name for anonymous components from other bundles.
```
Comment (smnandre) [inline src/TwigComponent/CHANGELOG.md]:
```suggestion
-  Extended support for the `index.html.twig` template fallback when resolving namespaced anonymous components.
```

I am not sure this adds _much_ precision, but it does better reflect the current implementation.

-> Call "Foo:Bar" and we will find the matching "Foo/Bar/index.html.twig" template

**Comment by smnandre** [comment]
Thank you for this nice first PR  @Phil23  🎉

**Review by smnandre — APPROVED** [review]

**Comment by Kocal** [comment]
Thank you @Phil23.
